### PR TITLE
Overhaul of train.py and adding Chesapeake CVPR trainer

### PR DESCRIPTION
**Pull request by calebrob6**
Some key points
- Changed how the configuration files are structured. See conf/defaults.yaml for the gist of this. There is now experiment.module and experiment.datamodule keys that get passed as kwargs to the LightningModule and LightningDataModule for a given task. This makes everything quite flexible. Because everything is a bit more flexible I could get rid of some nastyness in `train.py`.
- Fixed the tests accordingly
- `train.py` saves the final config that is used for the experiment to the experiment output directory
- Made a trainer for the Chesapeake CVPR dataset

And.... it is working well! Here is output from tensorboard showing val image, val mask, predictions:
image

Things to add in the nearish future:
- A way to train on a concat of different state train splits
- Augmentations

### CODE REVIEW

File: torchgeo/trainers/chesapeake.py [hunk 1]
```diff
@@ -109,12 +130,42 @@ def validation_step(  # type: ignore[override]
...
+            )
+            mask = batch["mask"][0].cpu().numpy()
+            pred = y_hat_hard[0].cpu().numpy()
+            fig, axs = plt.subplots(1, 3, figsize=(12, 4))
+            axs[0].imshow(img[:, :, :3])
+            axs[0].axis("off")
+            axs[1].imshow(mask, vmin=0, vmax=6, cmap=CMAP, interpolation="none")
+            axs[1].axis("off")
+            axs[2].imshow(pred, vmin=0, vmax=6, cmap=CMAP, interpolation="none")
+            axs[2].axis("off")
```
Comment (adamjstewart) [inline torchgeo/trainers/chesapeake.py:159]: Can we instead move the plotting logic to the `ChesapeakeCVPR` dataset class? That's how `RasterDataset` and `VectorDataset` are designed. Then you can just call `dataset.plot(sample)` and the code can be reused by anyone using the dataset in their own code.

Comment (calebrob6) [reply torchgeo/trainers/chesapeake.py:159]: `RasterDataset` plot takes a single Tensor as input, while this code is plotting (image, mask, predictions). Do you imagine multiple versions of plot?

Comment (adamjstewart) [reply torchgeo/trainers/chesapeake.py:159]: Ah, I think this is the first GeoDataset we have with both `image` and `mask`. We should change `plot` to take in a sample dict instead.

File: conf/chesapeake_cvpr.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+trainer:
```
Comment (adamjstewart) [inline conf/chesapeake_cvpr.yaml:1]: My biggest complaint about all this OmegaConf stuff is that there doesn't seem to be any documentation on what options are supported or what possible values they can take. There's no way to get a help message without argparse.

File: conf/task_defaults/chesapeake_cvpr.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    segmentation_model: "unet"
+    encoder_name: "resnet18"
+    encoder_weights: "imagenet"
+    encoder_output_stride: 16
+    learning_rate: 1e-3
+    learning_rate_schedule_patience: 2
+  datamodule:
+    batch_size: 32
+    num_workers: 4
+    train_state: "de"
```
Comment (adamjstewart) [inline conf/task_defaults/chesapeake_cvpr.yaml:14]: These files should really have line endings at the end...

File: torchgeo/trainers/chesapeake.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+        self.test_iou = IoU(num_classes=7)
+
+    def forward(self, x: Tensor) -> Any:  # type: ignore[override]
+        """Forward pass of the model."""
+        return self.model(x)
+
+    def training_step(  # type: ignore[override]
+        self, batch: Dict[str, Any], batch_idx: int
+    ) -> Tensor:
+        """Training step - reports average accuracy and average IoU."""
```
Comment (adamjstewart) [inline torchgeo/trainers/chesapeake.py:106]: Many of these docstrings aren't documenting args or return values, see #101

File: torchgeo/trainers/chesapeake.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+# https://github.com/pytorch/pytorch/pull/61045
+DataLoader.__module__ = "torch.utils.data"
+Module.__module__ = "torch.nn"
+
+CMAP = matplotlib.colors.ListedColormap(
+    [np.array(Chesapeake7.cmap[i + 1]) / 255.0 for i in range(6)]
+)
+
+
+class ChesapeakeCVPRSegmentationTask(LightningModule):
```
Comment (adamjstewart) [inline torchgeo/trainers/chesapeake.py:36]: These trainer modules desperately need unit tests at some point, our overall code coverage is plummeting, see #109

File: torchgeo/trainers/cyclone.py [hunk 6]
```diff
@@ -36,7 +36,7 @@ def config_task(self, kwargs: Dict[str, Any]) -> None:
         else:
             raise ValueError(f"Model type '{kwargs['model']}' is not valid.")
 
-    def __init__(self, **kwargs: Dict[str, Any]) -> None:
+    def __init__(self, **kwargs: Any) -> None:
```
Comment (adamjstewart) [inline torchgeo/trainers/cyclone.py:39]: Why the switch to `Any` here? In mypy language, `Any` basically means "skip type checks", so I'm trying to avoid it whenever possible. In the future, I would like to never use `Any` anywhere in the code base. For dicts like `**kwargs` I see no reason not to specify that it has to be a dict.

File: torchgeo/trainers/cyclone.py [hunk 7]
```diff
@@ -198,6 +199,20 @@ def setup(self, stage: Optional[str] = None) -> None:
...
+            transforms=self.custom_transform,
+            download=False,
+        )
+
+        self.all_test_dataset = TropicalCycloneWindEstimation(
+            self.root_dir,
+            split="test",
+            transforms=self.custom_transform,
+            download=False,
+        )
```
Comment (adamjstewart) [inline torchgeo/trainers/cyclone.py:214]: I get that it's nice to download a dataset in one function and instantiate it elsewhere, but doesn't that require us to rebuild the entire dataset index twice? Seems quite a bit slower.

File: torchgeo/trainers/cyclone.py [hunk 8]
```diff
@@ -222,7 +237,7 @@ def train_dataloader(self) -> DataLoader[Any]:
             batch_size=self.batch_size,
             num_workers=self.num_workers,
             shuffle=True,
-            pin_memory=True,
+            pin_memory=False,
```
Comment (adamjstewart) [inline torchgeo/trainers/cyclone.py:240]: I believe False is the default, so we could just remove this argument.

**Review by adamjstewart — DISMISSED** [review]
Gave things another review. My biggest complaint with this chunk of code is a lack of documentation and testing. I realize we're kind of busy trying to finish experiments before we can publish things, but we need to make sure that this code is actually well-tested and documented before we start making releases.

File: torchgeo/trainers/cyclone.py (hunk 6 above)
Comment (calebrob6) [reply torchgeo/trainers/cyclone.py:39]: Dict isn't needed when annotating **kwargs, see https://adamj.eu/tech/2021/05/11/python-type-hints-args-and-kwargs/ or https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values. As kwargs here can actually be Any I think it is legit

File: torchgeo/trainers/cyclone.py (hunk 7 above)
Comment (calebrob6) [reply torchgeo/trainers/cyclone.py:214]: According to pytorchlightning docs you shouldn't save state in `prepare_data`, but should download datasets there. You should setup the dataset objects in `setup`.

File: torchgeo/trainers/cyclone.py (hunk 8 above)
Comment (calebrob6) [reply torchgeo/trainers/cyclone.py:240]: Done

File: torchgeo/trainers/cyclone.py (hunk 6 above)
Comment (adamjstewart) [reply torchgeo/trainers/cyclone.py:39]: Oh cool! I guess ya learn something new every day

File: conf/chesapeake_cvpr.yaml (hunk 2 above)
Comment (calebrob6) [reply conf/chesapeake_cvpr.yaml:1]: This is discussed in much more detail here -- https://github.com/facebookresearch/hydra/issues/633.

For now I think comments in the yaml files are OK -- we can do more here. I think there will be few scenarios in which a user is trying to configure experiments without looking at the trainer code.

File: torchgeo/trainers/chesapeake.py (hunk 5 above)
Comment (calebrob6) [reply torchgeo/trainers/chesapeake.py:36]: Have you tried running a trainer :)?

We ran into a very obvious problem in "tested" code this morning: I wanted to run `python train.py` with the cyclone dataset. This works in my environment because the dataset is where I expect and this works in tests because we have fake data where we expect it. This did _not_ work in practice because other people don't have the dataset downloaded.

Comment (adamjstewart) [reply torchgeo/trainers/chesapeake.py:36]: At the moment `train.py` isn't "tested" code. In the long run, I want to have integration tests that use real data and may take several hours to run. These will only be run on release branches, not on main/PRs. Everything is set up to do this except someone needs to actually write the tests. The unit tests will only be for coverage and catching minor issues.

**Review by adamjstewart — APPROVED** [review]

**Comment by calebrob6** [comment]
@adamjstewart:
- Fixed line endings
- Added some docstrings. Many of these methods are overriding pytorch lightning methods so I'm not sure what we gain by annotating them further? How is this normally handled?
- Got rid of pin_memory=False, left shuffle=True/False to be explicit
- I'm not sure what to do about your other comments. I broadly agree with what you are saying but don't see the benefit of spending a lot of time getting things perfect until early Oct.

**Comment by adamjstewart** [comment]
> I'm not sure what we gain by annotating them further? How is this normally handled?

If you look at our current [API docs](https://torchgeo.readthedocs.io/en/latest/api/trainers.html), you'll see that many of these methods are undocumented and only mention the names of arguments, not what they mean. We should try adding [`autodoc_inherit_docstrings`](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_inherit_docstrings) and see if this will add the superclass docstrings. I don't want to document all inherited members, but I think this will inherit docstrings without documenting non-overridden members. 

> I'm not sure what to do about your other comments. I broadly agree with what you are saying but don't see the benefit of spending a lot of time getting things perfect until early Oct.

Yep, that's fine. None of these suggestions are requirements that need to be done now, just pointing them out so we can think about them. As soon as we submit the paper I plan on going back and adding a lot of unit tests and fixing bugs so we can safely release.
